### PR TITLE
Clusterer feasibility changes

### PR DESCRIPTION
**Pull request by sfkhalid**
With the changes in this branch, the clusterer will not pack molecules together that have conflicting PartitionRegions (meaning they have different floorplanning constraints).

#### Description
Most changes were made in /vpr/src/pack/cluster.cpp. The following changes were made in this file:
-In do_clustering, a new check was added for whether the block failed a floorplanning feasibility check.
-In start_new_cluster, the cluster is assigned an empty PartitionRegion, and that PartitionRegion gets updated to match the seed atom, if the seed atom has its own non-empty PartitionRegion
-In try_pack_molecule a new feasibility check for floorplanning feasibility was added
-A new routine, intersect_atom_cluster_part_regions is called in try_pack_molecule to check compatibility between the atom and cluster

#### Related Issue
This change adds functionality needed to address the changes wanted in issue #932

### CODE REVIEW

File: vpr/src/base/partition_region.cpp [hunk 1]
```diff
@@ -8,6 +8,14 @@ std::vector<Region> PartitionRegion::get_partition_region() {
     return partition_region;
 }
 
+bool PartitionRegion::empty() {
+    if (partition_region.size() == 0) {
```
Comment (sfkhalid) [inline vpr/src/base/partition_region.cpp:12]: style-wise - can just  if return partition_region.size() == 0

File: vpr/src/base/partition_region.h [hunk 2]
```diff
@@ -24,6 +24,11 @@ class PartitionRegion {
      */
     std::vector<Region> get_partition_region();
 
+    /**
+     * @brief Check if the PartitionRegion is empty (ie. holds no regions in its vector)
```
Comment (sfkhalid) [inline vpr/src/base/partition_region.h:28]: i.e. no constraint on this object

Comment (sfkhalid) [reply vpr/src/base/partition_region.h:28]: don't need to explain vector is empty (that's getting into implementation)

File: vpr/src/base/vpr_constraints.cpp [hunk 3]
```diff
@@ -47,3 +47,17 @@ std::vector<AtomBlockId> VprConstraints::get_part_atoms(PartitionId part_id) {
...
+
+int VprConstraints::get_num_partitions() {
+    return num_partitions;
+}
+
+void VprConstraints::set_num_partitions(int num_parts) {
+    num_partitions = num_parts;
+}
+
+PartitionRegion VprConstraints::get_partition_pr(PartitionId part_id) {
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints.cpp:59]: could add const since it doesn't change, but its ok since it's passed by value

File: vpr/src/base/vpr_constraints.h [hunk 4]
```diff
@@ -33,7 +33,7 @@ class VprConstraints {
     /**
      * @brief Return id of the partition the atom belongs to
      *
-     * If an atom is not in a partition (unconstrained), PartitionId::Invalid() is returned.
+     * If an atom is not in a partition (unconstrained), PartitionId::INVALID() is returned.
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints.h:36]: At the top of this file, add a short summary explaining the different classes connected to this one (Partition, PartitionRegion, Region)

File: vpr/src/base/vpr_constraints.h [hunk 5]
```diff
@@ -70,6 +89,12 @@ class VprConstraints {
      * Store all partitions
      */
     vtr::vector<PartitionId, Partition> partitions;
+
+    /**
+     * Store number of partitions - the number of partitions in the partitions vector.
+     * Used when echoing VprConstraints objects.
+     */
+    int num_partitions;
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints.h:97]: check if you need this or you can just use partitions.size()

File: vpr/src/base/vpr_constraints.h [hunk 6]
```diff
@@ -70,6 +89,12 @@ class VprConstraints {
      * Store all partitions
      */
     vtr::vector<PartitionId, Partition> partitions;
+
+    /**
+     * Store number of partitions - the number of partitions in the partitions vector.
+     * Used when echoing VprConstraints objects.
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints.h:95]: why is needed specifically for echoing? clarify reason

File: vpr/src/base/vpr_constraints_reader.cpp [hunk 7]
```diff
@@ -31,10 +31,16 @@ void load_vpr_constraints_file(const char* read_vpr_constraints_name) {
             read_vpr_constraints_name);
     }
 
-    VprConstraints constraints = reader.constraints_;
-    int num_parts = reader.num_partitions_;
+    //VprConstraints constraints = reader.constraints_;
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints_reader.cpp:34]: take out these commented lines, not needed anymore

File: vpr/src/base/vpr_constraints_reader.cpp [hunk 8]
```diff
@@ -31,10 +31,16 @@ void load_vpr_constraints_file(const char* read_vpr_constraints_name) {
...
+
+    //Update the floorplanning constraints in the floorplanning constraints context
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+    floorplanning_ctx.constraints = reader.constraints_;
+
+    VprConstraints ctx_constraints = floorplanning_ctx.constraints;
 
     if (getEchoEnabled() && isEchoFileEnabled(E_ECHO_VPR_CONSTRAINTS)) {
-        echo_constraints(getEchoFileName(E_ECHO_VPR_CONSTRAINTS), constraints, num_parts);
+        echo_constraints(getEchoFileName(E_ECHO_VPR_CONSTRAINTS), ctx_constraints, ctx_constraints.get_num_partitions());
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints_reader.cpp:44]: same thing, check if you can use partitions.size()

File: vpr/src/base/vpr_constraints.h (hunk 5 above)
Comment (sfkhalid) [reply vpr/src/base/vpr_constraints.h:97]: get_num_partitions can just directly return partitions.size() in its implementation

File: vpr/src/base/vpr_constraints_serializer.h [hunk 9]
```diff
@@ -295,6 +295,9 @@ class VprConstraintsSerializer final : public uxsd::VprConstraintsBase<VprConstr
     }
 
     virtual void finish_load() final {
+        //Pass the object the total number of partitions that were read in
+        //This number is used when echoing VprConstraints objects
+        constraints_.set_num_partitions(num_partitions_);
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints_serializer.h:300]: don't need this anymore, will already have access to num_partitions

File: vpr/src/base/vpr_context.h [hunk 10]
```diff
@@ -369,6 +370,17 @@ struct RoutingContext : public Context {
         cached_router_lookahead_;
 };
 
+/**
+ * @brief State relating to VPR's floorplanning constraints
+ *
+ * This should contain only data structures related to constraining blocks
+ * to certain regions on the chip.
+ */
```
Comment (sfkhalid) [inline vpr/src/base/vpr_context.h:378]: comment what the two data members are (vpr constraints can be brief, clb_constraints should have some more explanation)

Comment (sfkhalid) [reply vpr/src/base/vpr_context.h:378]: will be a bit faster in a vtr::vector because the IDs will be dense (every cluster will be in here, even if it just an empty PR)
there is a vtr::vector and vtr::vector_map - vector is probably enough in this case.

Comment (sfkhalid) [reply vpr/src/base/vpr_context.h:378]: vtr::vector is better for this case

File: vpr/src/base/vpr_types.h [hunk 11]
```diff
@@ -164,6 +164,7 @@ enum e_block_pack_status {
     BLK_PASSED,
     BLK_FAILED_FEASIBLE,
     BLK_FAILED_ROUTE,
+    BLK_FAILED_FLOORPLANNING,
```
Comment (sfkhalid) [inline vpr/src/base/vpr_types.h:167]: think about whether this enum can be in a lower header file. If it's only used in the clusterer, can have a cluster_utilities header file and put this enum in there. This can be a cleanup on a separate pull request

Comment (sfkhalid) [reply vpr/src/base/vpr_types.h:167]: in pack subdirectory - is there a cluster utilities type header file? check, could possibly put there

File: vpr/src/pack/cluster.cpp [hunk 12]
```diff
@@ -70,6 +70,8 @@
 #include "tatum/report/graphviz_dot_writer.hpp"
 #include "tatum/TimingReporter.hpp"
 
+#include "constraints_load.h"
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:73]: don't need this anymore

File: vpr/src/pack/cluster.cpp [hunk 13]
```diff
@@ -216,6 +218,10 @@ static enum e_block_pack_status try_place_atom_block_rec(const t_pb_graph_node*
                                                          int verbosity,
                                                          const int feasible_block_array_size);
 
+static enum e_block_pack_status intersect_atom_cluster_part_regions(const AtomBlockId blk_id,
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:221]: verbosity should have const too

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:221]: look at clustering context - if variables that don't change are passed all around, then you can pass around verbosity too. but if that is not the case, if they are grabbing control parameters from the clustering context then you can get verbosity from there too instead of passing it around to a bunch of different functions

File: vpr/src/pack/cluster.cpp [hunk 14]
```diff
@@ -633,6 +639,9 @@ std::map<t_logical_block_type_ptr, size_t> do_clustering(const t_packer_opts& pa
                                      next_molecule->pack_pattern->name, next_molecule->atom_block_ids.size());
                             VTR_LOG("\n");
                             fflush(stdout);
+                        } else if (block_pack_status == BLK_FAILED_FLOORPLANNING) {
+                            VTR_LOG("\tFloorplanning constraints not met: '%s' (%s)", blk_name.c_str(), blk_model->name);
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:643]: if you add the index (for using in debugger) of the atom, add for the other pack statuses as well. If not needed, don't add it though

File: vpr/src/pack/cluster.cpp [hunk 15]
```diff
@@ -1198,6 +1207,7 @@ static enum e_block_pack_status try_pack_molecule(t_cluster_placement_stats* clu
                                                   t_ext_pin_util max_external_pin_util) {
     int molecule_size, failed_location;
     int i;
+    int j;
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:1210]: just declare j in the loop, and can call it i_mol to specify it's a molecule

File: vpr/src/pack/cluster.cpp [hunk 16]
```diff
@@ -1233,6 +1243,24 @@ static enum e_block_pack_status try_pack_molecule(t_cluster_placement_stats* clu
         return BLK_FAILED_FEASIBLE;
     }
 
+    //for loop to check if cluster PartitionRegion intersects with atom PartitionRegion
+    for (j = 0; j < molecule_size; j++) {
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:1247]: int i_mol = 0, better style

File: vpr/src/pack/cluster.cpp [hunk 17]
```diff
@@ -1233,6 +1243,24 @@ static enum e_block_pack_status try_pack_molecule(t_cluster_placement_stats* clu
         return BLK_FAILED_FEASIBLE;
     }
 
+    //for loop to check if cluster PartitionRegion intersects with atom PartitionRegion
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:1246]: check if every atom in the molecule is legal in the cluster from a floorplanning perspective

File: vpr/src/pack/cluster.cpp (hunk 16 above)
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:1247]: need a temporary cluster_pr that you pass to the intersect routine. like the convention

File: vpr/src/base/echo_files.h [hunk 18]
```diff
@@ -12,6 +12,9 @@ enum e_echo_files {
     E_ECHO_PRE_PACKING_MOLECULES_AND_PATTERNS,
     E_ECHO_VPR_CONSTRAINTS,
 
+    //Packing
```
Comment (sfkhalid) [inline vpr/src/base/echo_files.h:15]: At top, file controls echo files, we have an enum to each one, search for in code to see what each one prints

Comment (sfkhalid) [reply vpr/src/base/echo_files.h:15]: say what echo files are, what they are used for - dumps that intended to be written and read by a developer, typically direct reps of data structures to make debugging easier

Comment (sfkhalid) [reply vpr/src/base/echo_files.h:15]: enum e_output_files - line 66 - see what these are in code and leave a brief comment, or make an issue saying this may just be same as other enum and we should just put them both together

Comment (sfkhalid) [reply vpr/src/base/echo_files.h:15]: set an issue to clean up these two enums (either just need to comment, or merge the two diff enums)
whole thing could be wrapped in a class

Comment (sfkhalid) [reply vpr/src/base/echo_files.h:15]: echo file name can maybe also take a default name (person can choose to pass it in - if not, default set to null), the set_output_filename would work fore everything

investigate - is there some other reason to have the second output enum? - look at the call sites for that

in the cpp file -> instead of the mallocs, recode using c++ strings and wrap it all in a class 

-> all the members of the class could be static members (don't have to turn into class, see if it's worth it) (make a member object of the echo class in something like the vpr_options file)

- might be better to not have class, either way clean up the code by getting rid of mallocs and use c++ strings

File: vpr/src/base/vpr_constraints_reader.cpp [hunk 19]
```diff
@@ -31,10 +31,13 @@ void load_vpr_constraints_file(const char* read_vpr_constraints_name) {
...
-    int num_parts = reader.num_partitions_;
+    //Update the floorplanning constraints in the floorplanning constraints context
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+    floorplanning_ctx.constraints = reader.constraints_;
+
+    VprConstraints ctx_constraints = floorplanning_ctx.constraints;
 
     if (getEchoEnabled() && isEchoFileEnabled(E_ECHO_VPR_CONSTRAINTS)) {
-        echo_constraints(getEchoFileName(E_ECHO_VPR_CONSTRAINTS), constraints, num_parts);
+        echo_constraints(getEchoFileName(E_ECHO_VPR_CONSTRAINTS), ctx_constraints, ctx_constraints.get_num_partitions());
```
Comment (sfkhalid) [inline vpr/src/base/vpr_constraints_reader.cpp:41]: don't need to pass third argument, since you're passing ctx_constraints, can just call ctx_constraints.get_num_partitions() inside echo_constraints - also pass ctx_constraints by const reference as a good convention

File: vpr/src/base/vpr_context.h [hunk 20]
```diff
@@ -369,6 +370,31 @@ struct RoutingContext : public Context {
...
+struct FloorplanningContext : public Context {
```
Comment (sfkhalid) [inline vpr/src/base/vpr_context.h:379]: mention in comments -> constraints are input into program, don't change. clustering_constraints will change if clusters change (basically talking about lifetime of the two variables)

Comment (sfkhalid) [reply vpr/src/base/vpr_context.h:379]: cluster_constraints computed during clustering

File: vpr/src/pack/output_clustering.cpp [hunk 21]
```diff
@@ -127,6 +130,28 @@ static void print_stats() {
     /* TODO: print more stats */
 }
 
+static void print_clustering_stats_header() {
+    VTR_LOG("----------   --------   ------------------------------------   --------------------------\n");
+    VTR_LOG("Block Type   # Blocks   Avg. # of input clocks and pins used   Avg. # of output pins used\n");
+    VTR_LOG("----------   --------   ------------------------------------   --------------------------\n");
+}
+
+static void print_clustering_stats(char* block_name, int num_block_type, float num_inputs_clocks, float num_outputs) {
```
Comment (sfkhalid) [inline vpr/src/pack/output_clustering.cpp:139]: Print out header: final clustering statistics

File: vpr/src/pack/cluster.cpp [hunk 22]
```diff
@@ -188,6 +188,11 @@ static t_pack_molecule* get_molecule_by_num_ext_inputs(const int ext_inps,
 static t_pack_molecule* get_free_molecule_with_most_ext_inputs_for_cluster(t_pb* cur_pb,
                                                                            t_cluster_placement_stats* cluster_placement_stats_ptr);
 
+static void print_pack_status(int num_clb,
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:191]: For any function added, make sure you have function level comment above it

File: vpr/src/pack/cluster.cpp [hunk 23]
```diff
@@ -788,6 +820,23 @@ std::map<t_logical_block_type_ptr, size_t> do_clustering(const t_packer_opts& pa
     return num_used_type_instances;
 }
 
+static void print_pack_status(int num_clb,
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:823]: add function comment

File: vpr/src/pack/cluster.cpp [hunk 24]
```diff
@@ -788,6 +820,23 @@ std::map<t_logical_block_type_ptr, size_t> do_clustering(const t_packer_opts& pa
...
+                              int tot_num_molecules,
+                              int num_molecules_processed,
+                              std::map<t_logical_block_type_ptr, size_t> clb_types) {
+    //print an update every 10 clusters
+    if (num_clb % 50 == 0) {
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:828]: calculate this as a percent of molecules, put a const float that will be the percentage of how often to print, will have num_molecules so can just multiply and see when to print

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:828]: might need another variable that is molecules since last print

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:828]: reset to 0 so you can have around every 4% message printing

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:828]: and pass in by reference

File: vpr/src/pack/cluster.cpp [hunk 25]
```diff
@@ -1331,6 +1402,12 @@ static enum e_block_pack_status try_pack_molecule(t_cluster_placement_stats* clu
                         }
                     }
 
+                    //update cluster PartitionRegion if atom with floorplanning constraints was added
+                    if (cluster_pr_needs_update) {
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:1406]: comment why you have a bool to update -> also change this to always return true and check that it still works, because it should

File: vpr/src/base/SetupGrid.cpp [hunk 26]
```diff
@@ -406,6 +406,8 @@ static DeviceGrid build_device_grid(const t_grid_def& grid_def, size_t grid_widt
         size_t incry = p.parse_formula(yspec.incr_expr, vars);
         size_t repeaty = p.parse_formula(yspec.repeat_expr, vars);
 
+        warn_out_of_range = false;
```
Comment (sfkhalid) [inline vpr/src/base/SetupGrid.cpp:409]: just pass in from called function instead of changing here

File: vpr/src/base/constraints_load.cpp [hunk 27]
```diff
@@ -67,16 +67,18 @@ void print_constraints(FILE* fp, VprConstraints constraints, int num_parts) {
...
+    int num_of_parts = constraints.get_num_partitions();
+
     fprintf(fp, "--------------------------------------------------------------\n");
     fprintf(fp, "Constraints\n");
     fprintf(fp, "--------------------------------------------------------------\n");
     fprintf(fp, "\n");
-    fprintf(fp, "\n Number of partitions is %d \n", num_parts);
-    print_constraints(fp, constraints, num_parts);
+    fprintf(fp, "\n Number of partitions is %d \n", num_of_parts);
+    print_constraints(fp, constraints, num_of_parts);
```
Comment (sfkhalid) [inline vpr/src/base/constraints_load.cpp:81]: clean up this argument too, don't have to pass in num_parts here either

Comment (sfkhalid) [reply vpr/src/base/constraints_load.cpp:81]: num_parts is a better name, if you can just get it in print_constraints, can just print Number of partition in print_constraints

File: vpr/src/pack/cluster.cpp [hunk 28]
```diff
@@ -572,6 +574,10 @@ std::map<t_logical_block_type_ptr, size_t> do_clustering(const t_packer_opts& pa
 
             VTR_LOGV(verbosity > 2, "Complex block %d:\n", num_clb);
 
+            /*this temp structure is used to store updates to the cluster's
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:577]: update comment, explain used to store updates to cluster's partitionregion as we try to add primitives to it. since some of the primitives might fail legality, we need this temp structure to store the legality

File: vpr/src/pack/cluster.cpp [hunk 29]
```diff
@@ -2633,6 +2641,31 @@ static void echo_clusters(char* filename) {
         }
     }
 
+    fprintf(fp, "\nCLUSTER CONSTRAINTS:\n");
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2644]: Change to cluster floorplanning constraints

File: vpr/src/base/constraints_load.cpp [hunk 30]
```diff
@@ -3,7 +3,7 @@
 static void print_region(FILE* fp, Region region);
```
Comment (sfkhalid) [inline vpr/src/base/constraints_load.cpp:3]: add to region class, (for each function, move to the class so it can print itself partition for partition, etc.)

Comment (sfkhalid) [reply vpr/src/base/constraints_load.cpp:3]: can be helper function in the class, or a global function after the class

File: vpr/src/pack/cluster.cpp [hunk 31]
```diff
@@ -2633,6 +2641,31 @@ static void echo_clusters(char* filename) {
...
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+
+    for (ClusterBlockId clb_id : cluster_ctx.clb_nlist.blocks()) {
+        std::vector<Region> reg = floorplanning_ctx.cluster_constraints[clb_id].get_partition_region();
+        if (reg.size() != 0) {
+            fprintf(fp, "\nRegions in Cluster %zu:\n", size_t(clb_id));
+            for (unsigned int i = 0; i < reg.size(); i++) {
+                vtr::Rect<int> rect = reg[i].get_region_rect();
+                int xmin = rect.xmin();
```
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2653]: make a print helper function for region, partitionregion, see if rect has a print function

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2653]: can also make a helper function for vtr::rect if there is not already one

**Comment by sfkhalid** [comment]
@litghost nightly has been taking a long time on this.

**Comment by litghost** [comment]
> @litghost nightly has been taking a long time on this.

Kokoro failed to start this job correct, and it will eventually die.  This is due to a kokoro bug.

**Comment by sfkhalid** [comment]
@litghost Does nightly seem to be running normally here? 
Also, I cleaned up my vtr working directory a lot before pushing commits to this branch again. I had a lot of files from previous reg test runs that were taking up space. I tried doing du -sh on the vtr root directory before and after the cleanup and it went from like 5G to 970M, so hopefully the working space issue doesn’t occur here, but please let me know if it does. Thanks!

**Comment by sfkhalid** [comment]
@litghost It says nightly failed but I can't see anything on the invocation details. Did kokoro fail to start it properly again, or were there some QoR failures?

**Comment by sfkhalid** [comment]
@litghost I started nightly again yesterday and it seems like it failed again. Are there some qor failures or other issues present?

**Comment by sfkhalid** [comment]
[qor_results_compare.xlsx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6045473/qor_results_compare.xlsx)
QoR results on this branch

**Comment by vaughnbetz** [comment]
Please attach the QoR data.  Note that the reviews by Sarah on her code were really joint reviews we did over zoom, and that she typed in.  I'm happy with the code.
After merge, should look at the vtr_util::rect_union to see if it is usable for some this (basically replace/merge some of your parition_regions with rect_union).
